src/sentry/incidents/endpoints/project_alert_rule_index.py

```diff
@@ -8,11 +10,122 @@
+                    dict(o.items() + [("name", _generate_rule_label(rule.project, obj, o))])
+                    for o in rule.data.get("conditions", [])
+                ],
+                "actions": [
+                    dict(o.items() + [("name", _generate_rule_label(rule.project, obj, o))])
+                    for o in rule.data.get("actions", [])
+                ],
+                "actionMatch": rule.data.get("action_match") or Rule.DEFAULT_ACTION_MATCH,
+                "frequency": rule.data.get("frequency") or Rule.DEFAULT_FREQUENCY,
+                "name": rule.label,
+                "dateCreated": rule.date_added,
+                "environment": environment.name if environment is not None else None,
+            }
+        else:
+            raise AssertionError("Invalid rule to serialize: %r" % type(rule))
+
+
+class ProjectCombinedRuleIndexEndpoint(ProjectEndpoint):
+    def get(self, request, project):
+        """
+        Fetches alert rules and legacy rules for an organization
+        """
+        if not features.has("organizations:incidents", project.organization, actor=request.user):
+            raise ResourceDoesNotExist
+
+        # TODO: How to implement cursor in results?
+        cursor = 0  # TODO: Get from request. Is it an integer? Date?
+        page_size = 25  # TODO: Get from request. Is it `limit`?
+
+        # alert_rule_queryset = AlertRule.objects.fetch_for_organization(project.organization).order_by('-date_added')[:page_size]
+        alert_rule_queryset = AlertRule.objects.fetch_for_project(project).order_by("-date_added")
+
+        legacy_rule_queryset = (
+            Rule.objects.filter(
+                project=project, status__in=[RuleStatus.ACTIVE, RuleStatus.INACTIVE]
+            )
+            .select_related("project")
+            .order_by("-date_added")
+        )
+
+        combined_rules = []
+        while len(combined_rules) < cursor + page_size and (
+            len(alert_rule_queryset) != 0 or len(legacy_rule_queryset) != 0
+        ):
+            alert_rule = alert_rule_queryset[0] if len(alert_rule_queryset) > 0 else None
+            legacy_rule = legacy_rule_queryset[0] if len(legacy_rule_queryset) > 0 else None
+            if alert_rule is not None and legacy_rule is not None:
+                if alert_rule.date_added > legacy_rule.date_added:
+                    next_rule = alert_rule
+                else:
+                    next_rule = legacy_rule
+            elif alert_rule is None:
+                next_rule = legacy_rule
+            elif legacy_rule is None:
+                next_rule = alert_rule
+
+            combined_rules.append((len(combined_rules), next_rule))
+            if isinstance(next_rule, AlertRule):
+                alert_rule_queryset = alert_rule_queryset[1:]
+            else:
+                legacy_rule_queryset = legacy_rule_queryset[1:]
+
+        combined_rules = combined_rules[cursor : cursor + page_size]
+
+        return self.paginate(
```

Comment:
Since you're already performing this pagination logic manually, you can just call `serialize` directly. You don't benefit from the logic much here, unless we write a custom paginator.